src/arns.lua

```diff
@@ -423,14 +424,106 @@ function arns.assertValidIncreaseUndername(record, qty, currentTimestamp)
 	return true
 end
 
+-- AUCTIONS
+
+--- Creates an auction for a given name
+--- @param name string The name of the auction
+--- @param timestamp number The timestamp to start the auction
+--- @param initiator string The address of the initiator of the auction
+--- @return Auction|nil The auction instance
+function arns.createAuction(name, timestamp, initiator)
+	if not arns.getRecord(name) then
+		error("Name is not registered. Auctions must be created for registered names.")
+	end
+	if arns.getAuction(name) then
+		error("Auction already exists for name")
+	end
+
+	local baseFee = demand.getFees()[#name]
+	local demandFactor = demand.getDemandFactor()
+	local auction = Auction:new(name, timestamp, demandFactor, baseFee, initiator, arns.calculateRegistrationFee)
+	NameRegistry.auctions[name] = auction
+	-- ensure the name is removed from the registry
+	arns.removeRecord(name)
+	return auction
+end
+
+function arns.getAuction(name)
+	return NameRegistry.auctions[name]
+end
+
+function arns.getAuctions()
+	return NameRegistry.auctions or {}
+end
+
+function arns.submitAuctionBid(name, bidAmount, bidder, timestamp, processId, type, years)
+	local auction = arns.getAuction(name)
+	if not auction then
+		error("Auction does not exist")
+	end
+
+	-- assert the bid is between auction start and end timestamps
+	if timestamp < auction.startTimestamp or timestamp > auction.endTimestamp then
+		-- TODO: we should likely clean up the auction if it is outside of the time range
+		error("Bid timestamp is outside of auction start and end timestamps")
+	end
+	local requiredBid = auction:getPriceForAuctionAtTimestamp(timestamp, type, years)
+	local requiredOrBidAmount = bidAmount or requiredBid
+	if requiredOrBidAmount < requiredBid then
+		error("Bid amount is less than the required bid of " .. requiredBid)
+	end
+
+	local finalBidAmount = math.min(requiredOrBidAmount, requiredBid)
+
+	-- check the balance of the bidder
+	if not utils.walletHasSufficientBalance(bidder, finalBidAmount) then
+		error("Insufficient balance")
+	end
+
+	local record = {
+		processId = processId,
+		startTimestamp = timestamp,
+		endTimestamp = type == "lease" and timestamp + constants.oneYearMs * years or nil,
+		undernameLimit = constants.DEFAULT_UNDERNAME_COUNT,
```

Comment:
If the record had previously increased its undername limit, is that lost in transition here?